File: _B_pendulum/python/hw2/pendulumAnimation.py
import matplotlib.pyplot as plt
import matplotlib.patches as mpatches
import numpy as np 
import pendulumParam as P

class pendulumAnimation:
    def __init__(self):
        self.flag_init = True  # Used to indicate initialization
        # Initialize a figure and axes object
        self.fig, self.ax = plt.subplots()
        # Initializes a list of objects (patches and lines)
        self.handle = []
        # Specify the x,y axis limits
        plt.axis([-3*P.ell, 3*P.ell, -0.1, 3*P.ell])
        # Draw line for the ground
        plt.plot([-2*P.ell, 2*P.ell], [0, 0], 'b--')
        # label axes
        plt.xlabel('z')

    def update(self, state):
        z = state.item(0)  # Horizontal position of cart, m
        theta = state.item(1)  # Angle of pendulum, rads
        # draw plot elements: cart, bob, rod
        self.draw_cart(z)
        self.draw_bob(z, theta)
        self.draw_rod(z, theta)
        self.ax.axis('equal')
        # Set initialization flag to False after first call
        if self.flag_init == True:
            self.flag_init = False

    def draw_cart(self, z):
        # specify bottom left corner of rectangle
        x = z-P.w/2.0
        y = P.gap
        corner = (x, y)
        # create rectangle on first call, update on subsequent calls
        if self.flag_init == True:
            # Create the Rectangle patch and append its handle
            # to the handle list
            self.handle.append(
                mpatches.Rectangle(corner, P.w, P.h,
                                   fc = 'blue', ec = 'black'))
            # Add the patch to the axes
            self.ax.add_patch(self.handle[0])
        else:
            self.handle[0].set_xy(corner)  # Update patch

    def draw_bob(self, z, theta):
        # specify center of circle
        x = z+(P.ell+P.radius)*np.sin(theta)
        y = P.gap+P.h+(P.ell+P.radius)*np.cos(theta)
        center = (x,y)
        # create circle on first call, update on subsequent calls
        if self.flag_init == True:
            # Create the CirclePolygon patch and append its handle
            # to the handle list
            self.handle.append(
                mpatches.CirclePolygon(center, radius=P.radius,
                    resolution=15, fc='limegreen', ec='black'))
            # Add the patch to the axes
            self.ax.add_patch(self.handle[1])
        else:
            self.handle[1]._xy = center

    def draw_rod(self, z, theta):
        # specify x-y points of the rod
        X = [z, z+P.ell*np.sin(theta)]
        Y = [P.gap+P.h, P.gap+P.h+P.ell*np.cos(theta)]
        # create rod on first call, update on subsequent calls
        if self.flag_init == True:
            # Create the line object and append its handle
            # to the handle list.
            line, =self.ax.plot(X, Y, lw=1, c='black')
            self.handle.append(line)
        else:
            self.handle[2].set_xdata(X)
            self.handle[2].set_ydata(Y)

File: _B_pendulum/unsupported_languages/matlab/hw2/pendulumAnimation.m
classdef pendulumAnimation
    properties
        cart_handle
        rod_handle
        bob_handle
        ell
        width
        height
        gap
        radius
    end
    methods
        %------constructor-----------
        function self = pendulumAnimation(P)
            self.ell = P.ell;
            self.width = P.w;
            self.height = P.h;
            self.gap = P.gap;
            self.radius = P.radius;
            
            figure(1), clf
            % draw ground track
            plot([-2*self.ell, 2*self.ell],[0,0],'k'); 
            hold on
            % initialize the cart, rod, and bob 
            self=self.draw_cart(P.z0);
            self=self.draw_rod(P.z0, P.theta0);
            self=self.draw_bob(P.z0, P.theta0);
            % Set the x,y axis limits
            axis([-3*self.ell, 3*self.ell, -0.1, 3*self.ell]); 
            xlabel('z'); % label x-axis
        end

        function self=update(self, state)
            % external method to update the animation
            z= state(1); % Horizontal position of cart, m
            theta = state(2);  % Angle of pendulum, rads
            self=self.draw_cart(z);
            self=self.draw_rod(z, theta);
            self=self.draw_bob(z, theta);
            drawnow
        end

        function self=draw_cart(self, z)
            % specify X-Y locations of corners of the cart
            X = [z-self.width/2, z+self.width/2,...
                 z+self.width/2, z-self.width/2];
            Y = [self.gap, self.gap,...
                 self.gap+self.height, self.gap+self.height];
            % plot or update cart
            if isempty(self.cart_handle)
                self.cart_handle = fill(X,Y,'b');
            else
                set(self.cart_handle,'XData',X);
            end
        end

        function self=draw_rod(self, z, theta)
            % specify X-Y locations of ends or rod
            X = [z, z+self.ell*sin(theta)]; 
            Y = [...
                self.gap+self.height,...
                self.gap + self.height + self.ell*cos(theta)...
                ]; 
            % plot or update rod
            if isempty(self.rod_handle)
                self.rod_handle = plot(X, Y, 'k');
            else
                set(self.rod_handle,'XData', X, 'YData', Y);
            end
        end

        function self=draw_bob(self, z, theta)
            th = 0:2*pi/10:2*pi;
            % center of bob
            center = [z + (self.ell+self.radius)*sin(theta) ...
                self.gap+self.height + ...
                (self.ell+self.radius)*cos(theta)];
            % points that define the bob
            X = center(1) + self.radius*cos(th);
            Y = center(2) + self.radius*sin(th);
            % plot or update bob
            if isempty(self.bob_handle)
                self.bob_handle = fill(X,Y,'g');
            else
                set(self.bob_handle,'XData',X);
                set(self.bob_handle,'YData',Y);
            end
        end 
    end
end